derive/lib.rs

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![cfg_attr(not(test), feature(proc_macro, proc_macro_lib))]
+
+#[cfg(not(test))] extern crate proc_macro;
+#[macro_use] extern crate quote;
+extern crate syn;
+
+#[cfg(not(test))]
+#[proc_macro_derive(HeapSizeOf)]
+pub fn expand_token_stream(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_string(&input.to_string()).parse().unwrap()
+}
+
+fn expand_string(input: &str) -> String {
+    let type_ = syn::parse_macro_input(input).unwrap();
+
+    let variant_code = match type_.body {
+        syn::Body::Struct(ref data) => {
+            vec![expand_variant(type_.ident.clone().into(), data)]
+        }
+        syn::Body::Enum(ref variants) => {
+            variants.iter().map(|variant| {
+                let path = syn::Path {
+                    global: false,
+                    segments: vec![type_.ident.clone().into(), variant.ident.clone().into()],
+                };
+                expand_variant(path, &variant.data)
+            }).collect()
+        }
+    };
+
+    let name = &type_.ident;
+    let (impl_generics, ty_generics, where_clause) = type_.generics.split_for_impl();
+    let mut where_clause = where_clause.clone();
+    for param in &type_.generics.ty_params {
+        where_clause.predicates.push(syn::WherePredicate::BoundPredicate(syn::WhereBoundPredicate {
+            bound_lifetimes: Vec::new(),
+            bounded_ty: syn::Ty::Path(None, param.ident.clone().into()),
+            bounds: vec![syn::TyParamBound::Trait(
+                syn::PolyTraitRef {
+                    bound_lifetimes: Vec::new(),
+                    trait_ref: syn::Path {
+                        global: true,
+                        segments: vec!["heapsize".into(), "HeapSizeOf".into()],
+                    }
+                },
+                syn::TraitBoundModifier::None
+            )],
+        }))
+    }
+
+    let tokens = quote! {
+        #type_
+
+        impl #impl_generics ::heapsize::HeapSizeOf for #name #ty_generics #where_clause {
+            #[inline]
+            fn heap_size_of_children(&self) -> usize {
+                let mut sum = 0;
+                match *self {
+                    #( #variant_code )*
+                }
+                sum
+            }
+        }
+    };
+
+    tokens.to_string()
+}
+
+fn expand_variant(path: syn::Path, variant: &syn::VariantData) -> quote::Tokens {
+    let mut open = quote::Tokens::new();
+    let mut close = quote::Tokens::new();
+    let fields = match *variant {
```

Comment:
This `match` statement seems like it would work better lower down. For example something like:

```rust
let ref fields: Vec<Ident> = variant.fields().iter().enumerate().map(...).collect();
let pat = match *variant {
    VariantData::Unit => return ...,
    VariantData::Struct(_) => quote! { #path { #(ref #fields),* } },
    VariantData::Tuple(_) => quote! { #path ( #(ref #fields),* ) },
};
quote! {
    #pat => { ... }
}
```